src/Modules/OrchardCore.Commerce/Services/ProductInventoryProvider.cs

```diff
@@ -0,0 +1,66 @@
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Extensions;
+using OrchardCore.Commerce.Inventory.Models;
+using OrchardCore.Commerce.Models;
+using OrchardCore.ContentManagement;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using YesSql;
+
+namespace OrchardCore.Commerce.Services;
+
+/// <summary>
+/// A product inventory provider that updates inventories of shopping cart items which have an <c>InventoryPart</c>.
+/// </summary>
+public class ProductInventoryProvider : IProductInventoryProvider
+{
+    private readonly IProductService _productService;
+    private readonly ISession _session;
+
+    public ProductInventoryProvider(IProductService productService, ISession session)
+    {
+        _productService = productService;
+        _session = session;
+    }
+
+    public int Order => 0;
+
+    public async Task<bool> IsApplicableAsync(IList<ShoppingCartItem> model)
+    {
+        var skuProducts = await _productService.GetSkuProductsAsync(model);
+
+        return model.All(item =>
+            skuProducts.TryGetValue(item.ProductSku, out var productPart) &&
+            productPart.ContentItem.Has<InventoryPart>());
+    }
+
+    public async Task<int> QueryInventoryAsync(string sku)
+    {
+        var inventoryPart = (await _productService.GetProductAsync(sku))?.As<InventoryPart>();
+        return inventoryPart != null ? (int)inventoryPart.Inventory.Value : 0;
```

Comment:
![image](https://user-images.githubusercontent.com/4007293/209410034-d7831fce-97fc-48c1-b228-1d6a02f3e085.png)
This will crash, because you can't safly cast `decimal?` into `int`. You should do `inventoryPart?.Inventory.Value is { } inventory ? (int)inventory : 0` instead.